File: backend/src/main/java/com/fpms/service/AuthService.java
package com.fpms.service;

import com.fpms.dto.request.*;
import com.fpms.dto.response.AuthResponse;
import com.fpms.dto.response.UserResponse;
import com.fpms.security.UserPrincipal;

public interface AuthService {

    /**
     * Đăng ký tài khoản người dùng mới với vai trò mặc định ROLE_CUSTOMER
     *
     * @param request thông tin đăng ký người dùng
     */
    void register(RegisterRequest request);

    /**
     * Đăng nhập hệ thống bằng Email/SĐT và Mật khẩu, cấp phát cặp JWT Token
     *
     * @param request thông tin đăng nhập (loginId, password)
     * @return AuthResponse chứa cặp token và thông tin user
     */
    AuthResponse login(LoginRequest request);

    /**
     * Đăng nhập hệ thống bằng Google ID Token, cấp phát cặp JWT Token
     *
     * @param request thông tin GoogleLoginRequest chứa idToken
     * @return AuthResponse chứa cặp token và thông tin user
     */
    AuthResponse loginWithGoogle(GoogleLoginRequest request);

    /**
     * Lấy thông tin tài khoản người dùng hiện tại từ phiên đăng nhập
     *
     * @param userPrincipal thông tin Principal của người dùng đang đăng nhập
     * @return UserResponse chứa thông tin tài khoản (ẩn mật khẩu)
     */
    UserResponse getCurrentUser(UserPrincipal userPrincipal);

    /**
     * Tiếp nhận yêu cầu quên mật khẩu, sinh mã OTP 6 số và gửi qua email
     *
     * @param request thông tin email người dùng
     */
    void forgotPassword(ForgotPasswordRequest request);

    /**
     * Xác thực tính hợp lệ của mã OTP trước khi cho phép thiết lập mật khẩu mới
     *
     * @param request thông tin email và mã OTP
     */
    void verifyOtp(VerifyOtpRequest request);

    /**
     * Đặt lại mật khẩu mới sau khi xác thực OTP thành công
     *
     * @param request thông tin email, mã OTP, mật khẩu mới và xác nhận mật khẩu
     */
    void resetPassword(ResetPasswordRequest request);
}

File: frontend/src/services/authService.ts
import api from './api';
import type { 
  ApiResponse, 
  RegisterRequest, 
  LoginRequest, 
  AuthResponse, 
  UserResponse,
  ForgotPasswordRequest,
  VerifyOtpRequest,
  ResetPasswordRequest 
} from '../types/auth';

/**
 * Tầng Dịch vụ Xác thực (Auth Service)
 * Đóng gói tất cả các lời gọi API liên quan đến tài khoản
 */
export const authService = {
  /**
   * Gọi API Đăng ký tài khoản khách hàng mới
   * Endpoint: POST /api/v1/auth/register
   */
  registerApi: async (data: RegisterRequest): Promise<ApiResponse<void>> => {
    const response = await api.post<ApiResponse<void>>('/auth/register', data);
    return response.data;
  },

  /**
   * Gọi API Đăng nhập hệ thống
   * Endpoint: POST /api/v1/auth/login
   */
  loginApi: async (data: LoginRequest): Promise<ApiResponse<AuthResponse>> => {
    const response = await api.post<ApiResponse<AuthResponse>>('/auth/login', data);
    return response.data;
  },

  /**
   * Gọi API Lấy thông tin tài khoản hiện tại
   * Endpoint: GET /api/v1/auth/me
   * Token tự động được gắn thông qua Axios Request Interceptor trong api.ts
   */
  getMeApi: async (): Promise<ApiResponse<UserResponse>> => {
    const response = await api.get<ApiResponse<UserResponse>>('/auth/me');
    return response.data;
  },

  /**
   * Gọi API Đăng nhập bằng Google ID Token
   * Endpoint: POST /api/v1/auth/google
   */
  loginWithGoogleApi: async (idToken: string): Promise<ApiResponse<AuthResponse>> => {
    const response = await api.post<ApiResponse<AuthResponse>>('/auth/google', { idToken });
    return response.data;
  },

  /**
   * Gọi API yêu cầu cấp mã OTP qua Email
   * Endpoint: POST /api/v1/auth/forgot-password
   */
  forgotPasswordApi: async (data: ForgotPasswordRequest): Promise<ApiResponse<void>> => {
    const response = await api.post<ApiResponse<void>>('/auth/forgot-password', data);
    return response.data;
  },

  /**
   * Gọi API xác thực mã OTP
   * Endpoint: POST /api/v1/auth/verify-otp
   */
  verifyOtpApi: async (data: VerifyOtpRequest): Promise<ApiResponse<void>> => {
    const response = await api.post<ApiResponse<void>>('/auth/verify-otp', data);
    return response.data;
  },

  /**
   * Gọi API đặt lại mật khẩu mới
   * Endpoint: POST /api/v1/auth/reset-password
   */
  resetPasswordApi: async (data: ResetPasswordRequest): Promise<ApiResponse<void>> => {
    const response = await api.post<ApiResponse<void>>('/auth/reset-password', data);
    return response.data;
  },

  /**
   * Xóa sạch token ở Client khi người dùng đăng xuất
   * Xóa ở cả localStorage (nếu có ghi nhớ) và sessionStorage (nếu không ghi nhớ)
   */
  logout: () => {
    localStorage.removeItem('accessToken');
    localStorage.removeItem('refreshToken');
    sessionStorage.removeItem('accessToken');
    sessionStorage.removeItem('refreshToken');
  },
};

